File: backend/app/models/transaction.py
import uuid
from sqlalchemy import Column, String, Numeric, Integer, DateTime, ForeignKey, func
from app.database import Base


class Transaction(Base):
    __tablename__ = "transactions"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    sender_merchant_id = Column(String, nullable=True, index=True)
    sender_user_id = Column(String, ForeignKey("users.id"), nullable=True, index=True)
    receiver_merchant_id = Column(String, nullable=True, index=True)
    receiver_user_id = Column(String, nullable=True, index=True)
    sender_bank_account_id = Column(String, nullable=True)
    receiver_bank_account_id = Column(String, nullable=True)
    amount = Column(Numeric(12, 2), nullable=False)
    currency = Column(String, default="USD")
    rail = Column(String, nullable=True)  # fednow, rtp, ach, card
    status = Column(String, default="pending")  # pending, processing, completed, failed, cancelled
    idempotency_key = Column(String, unique=True, nullable=False, index=True)
    reference_id = Column(String, nullable=True)
    failure_reason = Column(String, nullable=True)
    description = Column(String, nullable=True)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())

    # --- stablecoin / on-chain settlement scaffolding (not yet wired into logic) ---
    asset_code = Column(
        String, ForeignKey("assets.code"),
        nullable=False, default="USD", server_default="USD",
    )
    amount_base_units = Column(Numeric(38, 0), nullable=True)
    settlement_type = Column(
        String, nullable=False, default="offchain", server_default="offchain",
    )  # offchain | onchain
    settlement_network = Column(String, nullable=True)   # ethereum | solana | bnb
    onchain_tx_hash = Column(String, unique=True, nullable=True)
    onchain_status = Column(String, nullable=True)       # submitted|confirming|confirmed|failed|reorged
    confirmations = Column(Integer, nullable=True, default=0, server_default="0")
    partner = Column(String, nullable=True)              # 'zerohash'
    partner_transfer_id = Column(String, nullable=True, index=True)
    direction = Column(String, nullable=True)            # onramp|offramp|send|deposit|withdrawal

File: mobile_flutter/lib/models/transaction.dart
class Transaction {
  final String id;
  final String? senderMerchantId;
  final String? senderUserId;
  final String receiverMerchantId;
  final double amount;
  final String currency;
  final String? rail;
  final String status;
  final String idempotencyKey;
  final String? referenceId;
  final String? failureReason;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Transaction({
    required this.id,
    this.senderMerchantId,
    this.senderUserId,
    required this.receiverMerchantId,
    required this.amount,
    this.currency = 'USD',
    this.rail,
    required this.status,
    required this.idempotencyKey,
    this.referenceId,
    this.failureReason,
    this.createdAt,
    this.updatedAt,
  });

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'] as String,
      senderMerchantId: json['sender_merchant_id'] as String?,
      senderUserId: json['sender_user_id'] as String?,
      receiverMerchantId: json['receiver_merchant_id'] as String,
      amount: (json['amount'] is String)
          ? double.parse(json['amount'] as String)
          : (json['amount'] as num).toDouble(),
      currency: json['currency'] as String? ?? 'USD',
      rail: json['rail'] as String?,
      status: json['status'] as String,
      idempotencyKey: json['idempotency_key'] as String,
      referenceId: json['reference_id'] as String?,
      failureReason: json['failure_reason'] as String?,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'sender_merchant_id': senderMerchantId,
        'receiver_merchant_id': receiverMerchantId,
        'amount': amount.toString(),
        'currency': currency,
        'rail': rail,
        'status': status,
        'idempotency_key': idempotencyKey,
      };
}
